### PR TITLE
8280642: ObjectInputStream.readObject should throw InvalidClassException instead of IllegalAccessError

### PR DESCRIPTION
**Pull request by RogerRiggs**
During deserialization of a serialized data stream that contains a proxy descriptor with non-public interfaces
`java.io.ObjectInputStream` checks that the interfaces can be loaded from a single classloader in `ObjectInputStream.resolveProxyClass`.
If the interfaces cannot be loaded from a single classloader, an `IllegalAccessError` is thrown.
When `ObjectInputStream.readObject` encounters this case, it reflects an incompatibility
between the classloaders of the source of the serialized stream and the classloader being used for deserialization.
When a proxy object cannot be created from the interfaces, `ObjectInputStream.readObject` should catch
the `InvalidAccessError` and throw `InvalidObjectException` with the `InvalidAccessError` as the cause.
This allows the application to handle the exception consistently with other errors during deserialization.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8280642](https://bugs.openjdk.java.net/browse/JDK-8280642): ObjectInputStream.readObject should throw InvalidClassException instead of IllegalAccessError
 * [JDK-8280906](https://bugs.openjdk.java.net/browse/JDK-8280906): ObjectInputStream.readObject should throw InvalidClassException instead of IllegalAccessError (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7274/head:pull/7274` \
`$ git checkout pull/7274`

Update a local copy of the PR: \
`$ git checkout pull/7274` \
`$ git pull https://git.openjdk.java.net/jdk pull/7274/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7274`

View PR using the GUI difftool: \
`$ git pr show -t 7274`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7274.diff">https://git.openjdk.java.net/jdk/pull/7274.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/io/ObjectInputStream.java [hunk 1]
```diff
@@ -1995,6 +1995,10 @@ private ObjectStreamClass readProxyDesc(boolean unshared)
             }
         } catch (ClassNotFoundException ex) {
             resolveEx = ex;
+        } catch (IllegalAccessError err) {
+            IOException ice = new InvalidObjectException(err.getMessage());
```
Comment (naotoj) [inline src/java.base/share/classes/java/io/ObjectInputStream.java:1999]: Would the variable be `ioe`?

**Review by naotoj — APPROVED** [review]
Looks good.
In the CSR, I see some `InvalidAccessError` which should be `IllegalAccessError`.

**Comment by mlchung** [comment]
`ObjectInputStream::readObject` does not specify that `InvalidObjectException` may be thrown.   Have you considered throwing `InvalidClassException` be an alternative?    i.e. extend `InvalidClassException` to report failure in creating the proxy class during deserialization.

**Comment by RogerRiggs** [comment]
> `ObjectInputStream::readObject` does not specify that `InvalidObjectException` may be thrown. 
> Have you considered throwing `InvalidClassException` be an alternative? i.e. extend `InvalidClassException` to report failure in creating the proxy class during deserialization.

There are other checks related to the Proxy interfaces that throw InvalidObjectException.
InvalidClassException was a close second.
The throwing of `IllegalAccessException` is not described for `readObject` and `InvalidObjectException` is a subclass of `IOException` that acts as a catch-all.

**Comment by mlchung** [comment]
`readProxyDesc` throws a mix of `InvalidClassException` and `InvalidObjectException`.   I'm not close to the spec of this area and let others to comment which one would be appropriate for this case.    In general, wrapping IAE with an existing exception is a reasonable solution.

IMO,  it'd be helpful to clarify this in the javadoc and document this specific exception.

**Comment by RogerRiggs** [comment]
The CSR has been updated to reflect the use of InvalidClassException, please review.

**Review by mlchung — APPROVED** [review]
LGTM

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**